Client/Frontend/ContentBlocker/ContentBlockerHelper.swift

```diff
@@ -106,42 +106,56 @@ class ContentBlockerHelper {
         return EnabledState(rawValue: pref) ?? .onInPrivateBrowsing
     }
 
-    fileprivate func addActiveRulesToTab() {
-        guard let ruleStore = ruleStore else { return }
+    func addActiveRulesToTab() -> Deferred<[String]> {
+        let result = Deferred<[String]>()
+        guard let ruleStore = ruleStore else { return result }
         let rules = blocklistBasic + (blockingStrengthPref == .strict ? blocklistStrict : [])
         let enabledMode = enabledStatePref
         removeAllFromTab()
 
         func addRules() {
-            for name in rules {
+            let deferreds: [Deferred<String?>] = rules.map { name in
+                let result = Deferred<String?>()
                 ruleStore.lookUpContentRuleList(forIdentifier: name) { rule, error in
-                    self.addToTab(contentRuleList: rule, error: error)
+                    let ok = self.addToTab(contentRuleList: rule, error: error)
```

Comment:
The error is passed into this method. But then nothing is done with it besides having it displayed. Do that here instead. 
Same with checking if the rule is nil. You could also do it here.

Reply:
+1, it was used differently in an older patch